### ISSUE TITLE
Async dispatch in getInitialProps

### ISSUE THREAD

**Issue by JoaoFGuiomar** [bmealhouse/next-redux-saga#60]
Pretty sure I had this working before... but for some reason right now when I dispatch an action in getInitialProps it doesn't seem to wait for the saga to finish.

I am dispatching an action in `_app.js` `getInitialProps` before
`const pageProps = Component.getInitialProps ? await Component.getInitialProps(ctx, isLoggedIn) : {};` that should trigger a saga to get some customer info.

Inside another page's `getInitialProps` I'm dispatching an action that depends on some customer data:

```
SomeOtherPage.getInitialProps = async (ctx, isLoggedIn) => {
    // Fetch more data

    // The saga that fetches the customer data didn't finish yet.
    // making the following dispatch saga's to fail because they need customer data in the customer reducer.

    if (isLoggedIn) {
        ctx.store.dispatch({ type: actions.SOME_REQUEST });
        ctx.store.dispatch({ type: actions.ANOTHER_REQUEST_THAT_NEEDS_CUSTOMER_DATA });
    }

   (...)
```

My saga:
```
export function* watchSomeRequest() {
    yield takeLatest(SOME_REQUEST, initSomeRequest);
    yield takeLatest(ANOTHER_REQUEST_THAT_NEEDS_CUSTOMER_DATA, initAnotherRequest);
}
```

```
function* initAnotherRequest(action) {
    try {
        const params = action.payload;

        // customer data is null  at this point. (initial state of customer reducer)
        // probably because the customer request didn't finish before the next dispatch started
        const customerData = yield select(customerSelectors.getData); 

       (...)
```

What am I missing?

I was expecting that the first dispatch (the one in _app.js  getInitialProps that will fetch the customer data) would finish before calling the other page getInitialProps, and so, when that other page would dispatch  `ANOTHER_REQUEST_THAT_NEEDS_CUSTOMER_DATA` the customer info would be available 🤔

**Comment by bbortt** [comment]
We're indeed having problems with the root saga.. Please track #59 for more information.
Closed as duplicate.

**Comment by JoaoFGuiomar** [comment]
@bbortt since #59 closed as a `won't fix` any chance on a more detailed reply about this one?

Is there any way to make async dispatch of actions?

**Comment by bbortt** [comment]
Sure, will take a look at it right away. Gimme some minutes ;)

**Comment by bbortt** [comment]
> Pretty sure I had this working before...

Just pretty sure or sure? I can hardly imagine that this works every time! Think about a high ping to your backend and the (by design) asynchronous `saga` will never finish before the second `getInitialProps`. However I am aware that `saga` was running synchronous before version `2.0.0`, I think (not sure which exact version, but a major release for sure). Maybe you updated this dependency as well which could have caused your fault.

To your above question:
> Is there any way to make async dispatch of actions?

Asynchronously dispatched actions are extremely hard to control. @bmealhouse gave some more information about this [here](https://github.com/bmealhouse/next-redux-saga/issues/19#issuecomment-433702123):
>You can dispatch as many times as you'd like as long as you are inside a single saga. END will wait until all the calls to dispatch have resolved before shutting down.
When you are dispatching async actions between multiple sagas, this is where things start to break down. You would have to handle this situation using channels, however, I'd rather duplicate code vs. introduce the complexity of channels.

The problem of asynchronous actions is the following: It's hard (actually impossible) to know, which one will resolve the last. You would have to dispatch `END` only after this one. Therefore we queue up `END` right after the `getInitialProps`.

I personally never load user data in `saga`s. I do it in a synchronous call via whatever http client is in use. Because I want to check all security terms before displaying anything to the user.
If you insist on doing it in an asynchronous way, dispatch another event when all user data is available and start fetching data based on that.

Does this help?

**Comment by JoaoFGuiomar** [comment]
Sure does. Thanks!

My work around was to make an action that does all the things I need for that specific  case and just call that single action in `getInitialprops`.
This action will trigger a saga that will do a couple of fetches and get what I need.

👍

**Comment by bbortt** [comment]
Great! 🚀
